### PR TITLE
Adds Transformation class

### LINKED ISSUE #92: Transformations for varying scale parameters
### Feature description

When optimising parameters of varying scale, we overshoot acceptable parameter ranges during fitting (for non-bounded methods). Transforming / normalising the parameter space on input to the optimiser should solve this issue. An example of this issue is:

```python
import pybop
import numpy as np

parameter_set = pybop.ParameterSet("pybamm", "Chen2020")
model = pybop.lithium_ion.SPMe(parameter_set=parameter_set)

# Fitting parameters
parameters = [
    pybop.Parameter(
        "Positive electrode diffusivity [m2.s-1]",
        prior=pybop.Gaussian(3.43e-15, 1e-15),
        bounds=[1e-15, 5e-15],
    ),
]

sigma = 0.001
t_eval = np.arange(0, 900, 2)
values = model.predict(t_eval=t_eval)
corrupt_voltage = values["Terminal voltage [V]"].data + np.random.normal(
    0, sigma, len(t_eval)
)

dataset = [
    pybop.Dataset("Time [s]", t_eval),
    pybop.Dataset("Current function [A]", values["Current [A]"].data),
    pybop.Dataset("Terminal voltage [V]", corrupt_voltage),
]

# Generate problem, cost function, and optimisation class
problem = pybop.Problem(model, parameters, dataset)
cost = pybop.SumSquaredError(problem)
optim = pybop.Optimisation(cost, optimiser=pybop.GradientDescent)
optim.optimiser.set_learning_rate(0.025)

x, final_cost = optim.run()
print("Estimated parameters:", x)
```

Our implementation of Gradient Descent (#88) is unbounded and immediately select a candidate solution orders of magnitude higher than an acceptable range.

### Motivation

_No response_

### Possible implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by BradyPlanden**
# Description

Adds a transformation class for `BaseCost` classes. This class applies a transformation to the parameter space to a latent space for improved optimisation / sampling. 

## Issue reference
Fixes #92 

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [X] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [ ] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [ ] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

File: examples/scripts/spm_CMAES.py [hunk 1]
```diff
@@ -42,7 +44,7 @@
 # Generate problem, cost function, and optimisation class
 problem = pybop.FittingProblem(model, parameters, dataset, signal=signal)
 cost = pybop.SumSquaredError(problem)
-optim = pybop.CMAES(cost, max_iterations=100)
+optim = pybop.CMAES(cost, max_iterations=100, min_iterations=40)
```
Comment (BradyPlanden) [inline examples/scripts/spm_CMAES.py:47]:
```suggestion
optim = pybop.CMAES(cost, max_iterations=100)
```

File: pybop/optimisers/base_pints_optimiser.py [hunk 2]
```diff
@@ -10,7 +10,10 @@
 from pints import SequentialEvaluator as PintsSequentialEvaluator
 from pints import strfloat as PintsStrFloat
 
-from pybop import BaseOptimiser, Result
+from pybop import (
+    BaseOptimiser,
+    Result,
+)
```
Comment (BradyPlanden) [inline pybop/optimisers/base_pints_optimiser.py:16]:
```suggestion
from pybop import BaseOptimiser, Result
```

File: pybop/optimisers/base_pints_optimiser.py [hunk 3]
```diff
@@ -48,11 +51,9 @@ def __init__(self, cost, pints_optimiser, **optimiser_kwargs):
         self._evaluations = None
         self._iterations = None
 
-        # PyBOP doesn't currently support the PINTS transformation class
-        self._transformation = None
-
         self.pints_optimiser = pints_optimiser
         super().__init__(cost, **optimiser_kwargs)
+        self.f = self.cost
```
Comment (BradyPlanden) [inline pybop/optimisers/base_pints_optimiser.py:56]:
```suggestion
```

File: pybop/optimisers/base_pints_optimiser.py [hunk 4]
```diff
@@ -196,12 +197,12 @@ def _run(self):
         if self._needs_sensitivities:
 
             def f(x):
-                L, dl = self.cost.evaluateS1(x)
+                L, dl = self.f.evaluateS1(x)
```
Comment (BradyPlanden) [inline pybop/optimisers/base_pints_optimiser.py:200]:
```suggestion
                L, dl = self.cost.evaluateS1(x)
```

File: pybop/optimisers/base_pints_optimiser.py [hunk 5]
```diff
@@ -196,12 +197,12 @@ def _run(self):
...
                 return (L, dl) if self.minimising else (-L, -dl)
         else:
 
-            def f(x, grad=None):
-                return self.cost(x, grad) if self.minimising else -self.cost(x, grad)
+            def f(x):
+                return self.f(x) if self.minimising else -self.f(x)
```
Comment (BradyPlanden) [inline pybop/optimisers/base_pints_optimiser.py:205]:
```suggestion
                return self.cost(x) if self.minimising else -self.cost(x)
```

File: pybop/costs/base_cost.py [hunk 6]
```diff
@@ -24,34 +24,46 @@ class BaseCost:
...
             self._target = self.problem._target
             self.parameters.join(self.problem.parameters)
             self.n_outputs = self.problem.n_outputs
             self.signal = self.problem.signal
+            self.transformation = self.construct_transformation()
 
-    @property
-    def n_parameters(self):
-        return len(self.parameters)
+    def construct_transformation(self):
```
Comment (NicolaCourtier) [inline pybop/costs/base_cost.py:36]: Let's move this to the Parameters class.

**Review by NicolaCourtier — COMMENTED** [review]
Great work implementing the transformation classes! To retain modularity, let's make `transformation` a property of the Parameters class (as well as each Parameter) instead of the cost/optimiser.

File: pybop/parameters/parameter.py [hunk 7]
```diff
@@ -379,7 +409,7 @@ def initial_value(self) -> np.ndarray:
         initial_values = []
 
         for param in self.param.values():
-            if param.initial_value is None:
+            if param.initial_value is None and param.prior is not None:
```
Comment (NicolaCourtier) [inline pybop/parameters/parameter.py:412]: I think this needs a warning in case there is no initial_value or prior

**Review by NicolaCourtier — APPROVED** [review]
Thanks @BradyPlanden, transformations look like a useful addition!
